### ISSUE TITLE
 support expressions in queries that refer to other columns

### ISSUE THREAD

**Issue by jenpaulhus** [LMFDB/lmfdb#3580]
This is a request related to the group order search on HGCWA.  Currently, to search for linear expressions for the group order in terms of the genus, a large "or" statement is created for each genus in the database.  As @roed314 says in PR #3516, the better way to do this would be to,  "support expressions in the generated postgres query that refer to other columns".

**Comment by AndrewVSutherland** [comment]
Agreed.  For the benefit of those who might not be aware of this.  It is possible to perform queries that compare two columns by going directly to postgres, e.g. to get a list of the 11 rows in hgcwa_passports for which group_order = 84*(genus-1) you can do
```
from psycopg2.sql import SQL
from lmfdb import db

cur = db._execute(SQL("SELECT * from hgcwa_passports WHERE group_order = 84*(genus-1)"))

```
but there is currently no way to do this using the standard db backend interface.  A quick and dirty way to do this would be to simply accept a WHERE clause in place of the query parameter of lucky, count, and search, e.g. the above search could be specified as

`results = db.g2c_curves.search ("group_order = 84*(genus-1)")`

This would be trivial to implement and would provide a lot of functionality (e.g. if I wanted to find all the newforms that are the designated representative of their twist equivalence class I could just do

`results = db.mf_newforms.search("label=minimal_twist")`

What do you think @roed314 ?  this would only take a few lines of code in db backend and seems a lot easier to me than trying to adapt the current query dictionary interface to support more general postgres queries.

**Comment by roed314** [comment]
I'm hesitant about the specifics of your suggestion for a few reasons.  First, we need to do some kind of validation to prevent SQL injection attacks (we can't accept arbitrary strings).  Second, while this works fine for interactive use, it doesn't behave very nicely with the infrastructure we have set up for parsing inputs from input boxes, where we have search parsers that are called consecutively to create a dictionary.

Here's an alternative suggestion which I think will work better on both fronts.  We add a new special parsing command `"$raw"`, which takes a string.  This string is split up and every string of alphabetic characters is wrapped in double quotes (this will make Postgres treat them as a name, for example a column name).  Aside from these column names, the only other allowable characters in the string are arithmetic operations: `+-*/^`.  We'd need to think a bit further if we want to support congruence conditions in this way, but I think it already does a lot of what we need and isn't that bad to implement.  So, to do the search requested above, we would use
`db.g2c_curves.search({'group_order':{'$raw':'84*(genus-1)'}})`.

As a further feature, we could allow some pre-defined string substitutions to be specified either in the search parser or the backend parser, allowing one to use `g` rather than `genus`.  It would also be nice to support implicit multiplication.

Thoughts?

**Comment by edgarcosta** [comment]
I share David's concerns and I agree with his proposal.
